packages/protocol/contracts/common/MetaTransactionWallet.sol

```diff
@@ -0,0 +1,256 @@
+    nonce = nonce.add(1);
+    bytes memory returnData = _executeTransaction(destination, value, data);
+    emit MetaTransactionExecution(destination, value, data, nonce.sub(1), returnData);
+    return returnData;
+  }
+
+  /**
+   * @notice Executes a transaction on behalf of the signer.`
+   * @param destination The address to which the transaction is to be sent.
+   * @param value The CELO value to be sent with the transaction.
+   * @param data The data to be sent with the transaction.
+   * @return The return value of the transaction execution.
+   */
+  function executeTransaction(address destination, uint256 value, bytes memory data)
+    public
+    returns (bytes memory)
+  {
+    require(msg.sender == signer || msg.sender == owner(), "Invalid transaction sender");
+    bytes memory returnData = _executeTransaction(destination, value, data);
+    emit TransactionExecution(destination, value, data, returnData);
+    return returnData;
+  }
+
+  /**
+   * @notice Executes multiple transactions on behalf of the signer.`
+   * @param destinations The address to which each transaction is to be sent.
+   * @param values The CELO value to be sent with each transaction.
+   * @param data The concatenated data to be sent in each transaction.
+   * @param dataLengths The length of each transaction's data.
+   */
+  function executeTransactions(
+    address[] calldata destinations,
+    uint256[] calldata values,
+    bytes calldata data,
+    uint256[] calldata dataLengths
+  ) external {
+    require(
+      destinations.length == values.length && values.length == dataLengths.length,
+      "Input arrays must be same length"
+    );
+    uint256 dataPosition = 0;
+    for (uint256 i = 0; i < destinations.length; i++) {
+      executeTransaction(destinations[i], values[i], sliceData(data, dataPosition, dataLengths[i]));
+      dataPosition = dataPosition.add(dataLengths[i]);
+    }
+  }
+
+  /**
+   * @notice Returns a slice from a byte array.
+   * @param data The byte array.
+   * @param start The start index of the slice to take.
+   * @param length The length of the slice to take.
+   * @return A slice from a byte array.
+   */
+  function sliceData(bytes memory data, uint256 start, uint256 length)
+    internal
+    returns (bytes memory)
+  {
+    // When length == 0 bytes.slice does not seem to always return an empty byte array.
+    bytes memory sliced;
+    if (length > 0) {
+      sliced = data.slice(start, length);
+    }
+    return sliced;
+  }
+
+  /**
+   * @notice Executes a transaction on behalf of the signer.`
+   * @param destination The address to which the transaction is to be sent.
+   * @param value The CELO value to be sent with the transaction.
+   * @param data The data to be sent with the transaction.
+   * @return The return value of the transaction execution.
+   */
+  function _executeTransaction(address destination, uint256 value, bytes memory data)
```

Comment:
Share this code with `MultiSig` and (eventually) `Proposals` by breaking out into a library.